torch/_decomp/decompositions.py

```diff
@@ -1206,35 +1206,3 @@ def log_sigmoid_forward(self: Tensor) -> Tuple[Tensor, Tensor]:
     else:
         buffer = z
     return min - torch.log1p(z), buffer
-
-
-# The implementation matches torch.ops.aten.norm
-# torch.ops.aten.norm only supports numeric p, does not support Frobenius norm or nuclear norm
-# For 2-norm and -2 matrix norm, it doesn't compute the singular values, it just compute the norm the same as when p > 2.
-@register_decomposition([aten.norm.Scalar, aten.norm.ScalarOpt_dim])
```

Comment:
we still need a decomposition for torch.ops.aten.norm, I don't think you can just delete it.

Reply:
I reimplemented it in terms of `vector_norm`. Later on in this stack I would like to make `at::linalg_vector_norm` the main implementation for this operation, and change all the calls from `at::norm` to calls into `at::linalg_vector_norm`. This will need moving the MPS and sparse implementations of `norm` to `linalg_vector_norm`, but I think we can try. It could serve as practice to see how much we can deprecate an implementation that's used left right and centre in practice.

Reply:
The diff following that one is a bit large, so here's an explanation:

After decomposing `norm` into `vector_norm`, I also needed to add a decomposition to `vector_norm`. I was going to add also a decomposition for `matrix_norm` and `svd`, but then I realised that I didn't need to do that given that they are ImplicitComposite so I added them to the relevant lists.

After doing that, I had to fix `upcast_tensor`, as it was not upcasting the `dtype` arg to `double` when it was `float`.
Then I started getting that the default implementation for the backward of `renorm` was more accurate than the reference implementation. I was able to fix it for CUDA touching a couple things in `renorm_backward`, but it's still failing on CPU. I ended up bumping a bit the tolerance for this operation, but while investigating this, I realised that the call to `torch.pow(x, 2)` within the ref of `vector_norm` was not being decomposed into a call to `mul(x,x)`, so I asked Ed to look into this. The mismatch on CPU may come from this? It'd be weird to me as `pow` does implement this optimisation on CPU, so I don't think so, but I don't know where is it coming from really... :/

Reply:
This diff removes decomposition registration for `aten.norm` and this is not the state we want to be in. Each diff should not regress functionality.

Reply:
I added it back. There should be no regression

Reply:
Even more, I added support for the dtype arg.